src/Microsoft.OData.Edm/Validation/ValidationRules.cs

```diff
@@ -924,28 +924,62 @@ public static class ValidationRules
                     }
                 });
 
+        // Remove this in 8.x. It is superseded by EntityTypeInvalidKeyKeyDefinedInAncestor
         /// <summary>
         /// Validates that a key is not defined if there is already a key in the base type.
         /// </summary>
+        [Obsolete("Please use EntityTypeInvalidKeyKeyDefinedInAncestor")]
         public static readonly ValidationRule<IEdmEntityType> EntityTypeInvalidKeyKeyDefinedInBaseClass =
+            new ValidationRule<IEdmEntityType>(
+               (context, entityType) =>
+               {
+                   if (entityType.BaseType != null &&
+                       entityType.DeclaredKey != null &&
+                       entityType.BaseType.TypeKind == EdmTypeKind.Entity &&
+                       entityType.BaseEntityType().DeclaredKey != null)
+                   {
+                       context.AddError(
+                       entityType.Location(),
+                       EdmErrorCode.InvalidKey,
+                       Strings.EdmModel_Validator_Semantic_InvalidKeyKeyDefinedInBaseClass(entityType.Name, entityType.BaseEntityType().Name));
+                   }
+               });
+
+        /// <summary>
+        /// Validates that a key is not more than once in a type hierarchy
+        /// </summary>
+        public static readonly ValidationRule<IEdmEntityType> EntityTypeInvalidKeyKeyDefinedInAncestor =
             new ValidationRule<IEdmEntityType>(
                 (context, entityType) =>
                 {
-                    if (entityType.BaseType != null &&
-                        entityType.DeclaredKey != null &&
-                        entityType.BaseType.TypeKind == EdmTypeKind.Entity &&
-                        entityType.BaseEntityType().DeclaredKey != null)
+                    bool foundKey = entityType.DeclaredKey != null;
+                    IEdmEntityType baseType = entityType;
+                    while ((baseType = baseType.BaseEntityType()) != null)

```

Comment:
Will we have already checked for cycles in the inheritance tree before using this function?

Reply:
That's a really good question --  turns out that yes, we do have code in Validate() to protect against circular references in type hierarchies.